[Windows] Create service utils

richl-dd opened the pull request:
### What does this PR do?
This PR creates a file that exposes utilities to interact with Windows services via the Service Control Manager (SCM). It wraps `OpenSCManager` and `OpenService` to allow the caller to specify the `desiredAccess` rights (as opposed to the Go package that uses `ALL_ACCESS`). This will allow agent code to use least-privilege accesses when working with Windows services. 

### Motivation
The motivation is to enable least-privileges SCM interactions from one place in the agent repo.

### Additional Notes
- The `ListDependentServices` function will list first, second, third, ... level dependencies
- It doesn't seem like Go is planning to make this change upstream:
  - https://github.com/golang/go/issues/48777
- Some of the test code helper functions are copy/pasted from: `go\pkg\mod\golang.org\x\sys@v0.4.0\windows\svc\mgr\mgr_test.go`. Assuming this is fine, because in the above issue the answerer suggested copying code directly from the package
- There is an open issue to add `enumDependentServices` to the Go package, when that is merged, we should update our code to use that function
  - https://github.com/golang/go/issues/56766 

### Possible Drawbacks / Trade-offs
- None I was able to identify at this time.

### Describe how to test/QA your changes
- There is some coverage in the unit tests:
  - https://github.com/DataDog/datadog-agent/pkg/util/winutil/service_test.go
- The current kitchen tests install and run the agent, so those should be sufficient
- To manually test this, installing the agent should also be sufficient

### Reviewer's Checklist
- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [x] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] Changed code has automated tests for its functionality.
- [x] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [x] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [x] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [x] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [x] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [x] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

richl-dd committed:
[Windows] WIP: consolidate Service Control Manager into common utils

richl-dd committed:
[WIP] use winutil service functions in controlsvc for agent

richl-dd committed:
[WIP] Fix enumDependentServices and add tests

- enumDependentServices now returns an empty slice and no error if there
  are no dependent services

- add tests that use least priviledge to:
  - open SC manager
  - create a service
  - open a service handle
  - list dependent services
  - delete services

clarkb7 commented on pkg/util/winutil/service.go:
```diff
@@ -0,0 +1,232 @@
+	return &mgr.Service{Name: serviceName, Handle: h}, err
+}
+
+func StartService(serviceName string) error {
```
Add a comment saying that this function does not block until the service is started. can link to [msdn ](https://learn.microsoft.com/en-us/windows/win32/api/winsvc/nf-winsvc-startservicea#remarks)for more info.

richl-dd replied:
Good call, added.

clarkb7 commented on pkg/util/winutil/service.go:
```diff
@@ -0,0 +1,232 @@
+	return nil
+}
+
+func ControlService(serviceName string, command svc.Cmd, to svc.State, desiredAccess uint32, timeout int64) error {
```
add a comment describing the wait behavior and the timeout parameter

richl-dd replied:
Good call, added.

clarkb7 commented on pkg/util/winutil/service.go:
```diff
@@ -0,0 +1,232 @@
+
+	deps, err := ListDependentServices(serviceName, windows.SERVICE_ACTIVE)
+	if err != nil {
+		return fmt.Errorf("could not list dependent services")
```
this loses the error reason. make sure to pass it on.

clarkb7 commented on pkg/util/winutil/service.go:
```diff
@@ -0,0 +1,232 @@
+	return err
+}
+
+// when Go has their version, replace ours with the upstream
```
link to PR/issue

richl-dd replied:
All set

clarkb7 commented on pkg/util/winutil/service.go:
```diff
@@ -0,0 +1,232 @@
+
+		ess.ServiceName = ess.ServiceName - uint64(baseptr)
+		ess.DisplayName = ess.DisplayName - uint64(baseptr)
+		ss := EnumServiceStatus{serviceName: ConvertWindowsString(servicearray[ess.ServiceName:]),
```
can you not pass `ess.ServiceName` instead of using the offset with `servicearray[ess.ServiceName:]`?

richl-dd committed:
[WIP] Test first and second level dependents in TestListDependentServices

richl-dd committed:
[WIP] Address comments from initial review

richl-dd committed:
Add copywrite and build tags to test file

- Add another test verification step for the order in which services are
  returned in ListDependentServices

richl-dd committed:
[Windows] Use internal doStopService

iglendd commented on pkg/util/winutil/service.go:
```diff
@@ -0,0 +1,241 @@
+}
+
+func StopService(serviceName string) error {
+
```
I wonder if you can create a single service array list ordered from non-depended to more depended services. I am not 100% sure but don't you need to collect dependents of the dependents (and order them appropriately)?

richl-dd replied:
> I wonder if you can create a single service array list ordered from non-depended to more depended services.

That is what `ListDependentServices` returns for us. It wraps `enumDependentServices`, which returns a slice of all dependent services in "reverse start order". From what I can tell, no, we don't need to collect dependents of dependents - they are all returned at once.

[Docs](https://learn.microsoft.com/en-us/windows/win32/api/winsvc/nf-winsvc-enumdependentservicesa)
[Example code](https://learn.microsoft.com/en-us/windows/win32/services/stopping-a-service)
[Unit tests](https://github.com/DataDog/datadog-agent/blob/richl/windows-service-utils/pkg/util/winutil/service_test.go#L91)

iglendd replied:
Interesting. I did not realize it. The very minor suggestion from previous is still possible but it is a nitpicking. All services can be stuffed into array and looped the same way. Would save few lines at least.

richl-dd replied:
Ah, I see what you're saying. I think I'd prefer to keep it as-is, so that its clear to the reader that first all the dependent services are being stopped, and then the desired service is stopped.

iglendd commented on pkg/util/winutil/service.go:
```diff
@@ -0,0 +1,241 @@
+// Start serviceName via SCM
+// Does not block until service is started
+// https://learn.microsoft.com/en-us/windows/win32/api/winsvc/nf-winsvc-startservicea#remarks
+func StartService(serviceName string) error {
```
I wonder if StartService cound be writtent in terms/logic of ControlService (with exception of Start and not control). But to have params timeout which by default could be 0.

richl-dd replied:
Could you elaborate on that? I think I see what you're saying, but not 100% clear.

iglendd replied:
What I was saying that StartService can call ControlService and get benefit of timeout control. The only  enchancement would require is to have a bit to Start vs Control svc. Event that can be inferred from the req state.

richl-dd replied:
Ah okay, so there isn't a start control command, so I don't think this is an option.

iglendd replied:
It is just a though and you do not to bother further or reply but imaginge if you want to add timeout to the StartService as well. You would endup implementing logic of what you have in ControlService minus a one line single API difference.  I was nitpicking of possibility of re-using ControlService-like functionality. Just to explain my thoughs above, May be it should be doControlService instead to keep external API correct. But again, it is just a thought.

iglendd commented on pkg/util/winutil/service.go:
```diff
@@ -0,0 +1,241 @@
+// https://learn.microsoft.com/en-us/windows/win32/api/winsvc/nf-winsvc-startservicea#remarks
+func StartService(serviceName string) error {
+
+	m, err := OpenSCManager(windows.SC_MANAGER_CONNECT)
+	if err != nil {
+		return err
+	}
+	defer m.Disconnect()
+
+	s, err := OpenService(m, serviceName, windows.SERVICE_START)
+	if err != nil {
+		return fmt.Errorf("could not open service: %v", err)
+	}
+	defer s.Close()
```
This section is used in many places, I wonder if it worth it to factored into separate function. It should save lot of lines

iglendd replied:
exactly!

richl-dd replied:
I thought about it a little more - I'd prefer to keep it as-is for now because it isn't clear from the docs if we can safely disconnect the SCM handle and still use the service handle. Happy to revisit, but for this PR, I'm going to leave it.

cc @clarkb7 @derekwbrown

iglendd commented on pkg/util/winutil/service.go:
```diff
@@ -0,0 +1,241 @@
+	timesup := time.Now().Add(time.Duration(timeout) * time.Second)
+	for status.State != to {
+		if time.Now().After(timesup) {
+			return fmt.Errorf("timeout waiting for service to go to state=%d", to)
```
I think it worth to dump status.State here to know on what state we have timeout ... and service name

richl-dd replied:
Yep, good catch, will update.

iglendd commented on pkg/util/winutil/service.go:
```diff
@@ -0,0 +1,241 @@
+	if err == error(windows.ERROR_SUCCESS) {
+		err = nil
+		return
+	} else {
```
There no need for else here after return above, right

richl-dd replied:
Yep, good catch, will update.

iglendd commented on pkg/util/winutil/service.go:
```diff
@@ -0,0 +1,241 @@
+
+	s, err := OpenService(m, serviceName, windows.SERVICE_START)
+	if err != nil {
+		return fmt.Errorf("could not open service: %v", err)
```
Should we add service name here and in a number of other places?

richl-dd replied:
Yes, probably 🙃 .. I'll do a pass on this.

iglendd reviewed (commented):
I wonder if we have other places in Agent where SCM is used and we probably can switch to this new code.

richl-dd commented on pkg/util/winutil/service.go:
```diff
@@ -0,0 +1,238 @@
+	for _, dep := range deps {
+		err = StopService(dep.serviceName)
+		if err != nil {
+			return fmt.Errorf("could not stop service")
```
Fixed, thanks.

richl-dd committed:
[WIP] Address some review comments

- Include service name in more logging statements
- Remove unneeded "else"

richl-dd committed:
[Windows] use more informative names

- as I was re-reading, I found that the "m" and "s" variables were confusing to
  keep track of, so I updated them to better reflect what they held

richl-dd committed:
[Windows] Add args to StartService

- include service name in logging

richl-dd committed:
[Windows] minor updates prior to merge

- use Go return conventions
